### PR TITLE
VMO-5090 Validate expressions

### PR DESCRIPTION
**Pull request by object-required**

### CODE REVIEW

File: package.json
Comment (safydy-r) [inline package.json:27]: @object-required , I think we may have a version conflict here. In fact,
- you branch flow-runner from `master`, right ?
- I already merged flow-runner rc3 into master https://github.com/FLOIP/flow-runner/pull/67, I didn't know we will have flow-runner for your ticket
- Which means, if we run `yarn install` in flow-builder, it will consume flow-runner `rc3`. And will will have a TS error about `Property 'config' does not exist on type 'IBlockExit'.` in flow-builder.

Trying to resolve this could be a bit tricky at this stage, so I'd like to suggest we consume `rc4` in your tickets, meaning:
- base flow-runner from this branch https://github.com/FLOIP/flow-runner/pull/68  instead of master
- base flow-builder from this branch https://github.com/FLOIP/flow-builder/pull/219 instead of master

Anyway, those rc4 PRs are already approved, moving them to QA

File: src/components/interaction-designer/toolbar/BlockErrorsExpandable.vue
Comment (safydy-r) [inline src/components/interaction-designer/toolbar/BlockErrorsExpandable.vue:116]: can we add a separator `/` between `${uuid}` & `${ajvError.dataPath}`?

File: src/components/interaction-designer/toolbar/ErrorNotifications.vue
Comment (safydy-r) [inline src/components/interaction-designer/toolbar/ErrorNotifications.vue:135]: Let's make an `early return` pattern for this. Pseudo code:
```
if(hasBlockValidationError) {
   return true
} else if (hasResourceValidationError) {
   return true
}

// continue other instructions 
return hasChoiceValidationError
```

This will avoid us to always run the `hasChoiceValidationError = Object.values(block.config.choices)
          .reduce((result, uuid) => (result === true || has(this.resourceValidationStatusesForActiveFlow, `resource/${uuid}`)), false)`

File: src/components/interaction-designer/toolbar/ErrorNotifications.vue
Comment (safydy-r) [inline src/components/interaction-designer/toolbar/ErrorNotifications.vue:130]: Could you explain a bit why we need `hasChoiceValidationError` ?

File: src/components/interaction-designer/toolbar/ErrorNotifications.vue
Comment (safydy-r) [inline src/components/interaction-designer/toolbar/ErrorNotifications.vue:146]: Good catch !

File: src/router/helpers.ts
Comment (safydy-r) [inline src/router/helpers.ts:5]: Why this changes ? this may affect line like https://github.com/FLOIP/flow-builder/pull/221/files#diff-9754383fb88bba0a3f438d1811b59308b89ee44c5c5c5d202b0409d83b8a5e89R16

Comment (safydy-r) [inline src/router/helpers.ts:5]: I'm curious, why this change ?

Comment (safydy-r) [reply src/router/helpers.ts:5]: If this is necessary, maybe we need to consider other refs like this https://github.com/FLOIP/flow-builder/pull/221/files#diff-9754383fb88bba0a3f438d1811b59308b89ee44c5c5c5d202b0409d83b8a5e89R16, right ?

File: src/store/flow/block.ts
Comment (safydy-r) [inline src/store/flow/block.ts:85]: Thanks for address this TODO, could we add a JsDoc mentioning we'd like to `make nested assignment reactive` ?
(that will explain the chunk usage)

File: src/store/validation/validationHelpers.ts
Comment (safydy-r) [inline src/store/validation/validationHelpers.ts:35]: I'm glad this is working !

File: src/components/interaction-designer/toolbar/BlockErrorsExpandable.vue
Comment (object-required) [reply src/components/interaction-designer/toolbar/BlockErrorsExpandable.vue:116]: The template looks inconsistent, but `dataPath` already contains a slash

Comment (safydy-r) [reply src/components/interaction-designer/toolbar/BlockErrorsExpandable.vue:116]: ah ok, thanks for confirming

File: src/components/interaction-designer/toolbar/ErrorNotifications.vue
Comment (object-required) [reply src/components/interaction-designer/toolbar/ErrorNotifications.vue:135]: I tried using short-circuit to optimize execution

File: src/components/interaction-designer/toolbar/ErrorNotifications.vue
Comment (object-required) [reply src/components/interaction-designer/toolbar/ErrorNotifications.vue:130]: No particular reason apart from I needed some distinction between content-resources and choice-option-resources during the development.

I removed it in the current version.

File: package.json
Comment (safydy-r) [reply package.json:27]: @object-required , After QA, and you merged flow-runner, make sure you use semantic versioning from npm here `^1.0.0-rc4` ;)

**Review by safydy-r — APPROVED** [review]
@object-required , Well done ! 👍 💯 I'm pre-approving to unblock you moving this forward.
I only have 02 requests:
1- There are TS errors when we serve the app, see the Travis-CI last build for details https://app.travis-ci.com/github/FLOIP/flow-builder/builds/249913708#L775 . Make sure the build passes !
2- About consuming flow-runner from npm once this is QA-ed

**Comment by pascaleli** [comment]
@object-required 

We are unable to test VMO-5090 & VMO-5731 appropriately , due to the fact that Expression evaluator for all the listed blocks in the ticket is broken. 
As shown in the picture below , I can enter wrong expressions or leave the Expression text box without getting any validation issue . 

Can you please fix this issue in your ticket or Inform me so I create a new ticket for it to help unblock testing your two tickets  


image

**Comment by safydy-r** [comment]
@object-required , I left a [comment](https://viamoinc.atlassian.net/browse/VMO-5090?focusedCommentId=19589) in the ticket to address @pascaleli 's above remark. Sorry for not spotting this earlier.

**Review by pascaleli — APPROVED** [review]

**Pull request merged**
